ops/testing.py

```diff
@@ -2586,6 +2605,128 @@ class NonAbsolutePathError(Exception):
     """
 
 
+class ContainerFilesystem(Protocol):
+    """The subset of container operations that are filesystem-related.
+
+    Don't instantiate this directly; use :meth:`Harness.get_container_filesystem`.
+    """
+
+    @abstractmethod
```

Comment:
IIRC, with Protocols, you don't have to explicitly implement the methods, nor make them abstract, since you're not supposed to subclass a Protocol.

Typically you do
```
class Foo(Protocol):
    def bar(self, a: str, b:int, **) -> MyReturnType: ...
```

No docstrings, no raise NotImplementedError... just an Ellipsis will do.
Again, nothing will subclass this, so the docstrings are just 'wasted'. What a protocol defines is the 'api' of an object.

Reply:
(if you want to preserve docstrings on your proxy object, you should 'copy them' over from `model.Container` 
like:
```python
def _copy_docstrings(source_cls: Any):
    """Copy the docstrings from source_cls to target_cls.

    Use this as:
      @_copy_docstrings(source_class)
      class TargetClass:

    And for any public method that exists on both classes, it will copy the
    __doc__ for that method.
    """
    def decorator(target_cls: Any):
        for meth_name, _ in target_cls.__dict__.items():
            if meth_name.startswith('_'):
                continue
            source_method = source_cls.__dict__.get(meth_name)
            if source_method is not None and source_method.__doc__:
                target_cls.__dict__[meth_name].__doc__ = source_method.__doc__
        return target_cls
    return decorator
```

(from `ops.testing`)

Reply:
@benhoyt mentioned that in the [the Protocol PEP 544](https://peps.python.org/pep-0544/) examples, the `NotImplementedError` is used in the abstract methods.
Also, here's a collapsed conversation about using `_copy_docstring` https://github.com/canonical/operator/pull/933#discussion_r1199858265

Reply:
yeah but do they need to be abstractmethods in the first place?

Reply:
I think in our use case, with or without `abstractmethod` are interchangeable.